### DISCUSSION TITLE
request for hints concerning specific `make` and Jenkins options

### DISCUSSION THREAD

**Discussion by bansp** [TEIC/TEI#2628] [Q&A] [answered]
Hi all,

This is directed at experts in the intricacies of the options provided by the Guidelines Makefile and possibly also Jenkins.
For the purpose of testing proposals developed within the LingSIG, I would love to identify the options of make that would give us:

1. `P5/release/xml/tei/odd/p5.nvdl`, in order to substitute it in the PIs that identify schemas in the parts of the Guidelines that we happen to be working on
2. the HTML versions of the Guidelines (including the specs and possibly, but not necessarily for now, the exemplars)
3. possibly some early testing that would signal issues before our suggestions are submitted to the Council

At the same time, I would like to _avoid_:

a. generating Guidelines other than "en"
b. all the zipping -- we don't need archives, we just need the artefacts (and only for "en")

Currently, I use the following for Jenkins:
`make VCS=git INJENKINS=true XSL=${JENKINS_HOME}/jobs/Stylesheets-dev/lastSuccessful/archive/dist/xml/tei/stylesheet clean dist` -- and that produces the .nvdl and the Guidelines, along with a _heap of stuff that I don't need at all_, due to which the last Jenkins run took over 1.5h... that's a bit mucho, I'd prefer something like, errm, 5 minutes, which should in theory be enough for just (1) and (2) above, without the (a) and (b).

I have tried a few combinations such as, e.g., `clean validate exemplars test html-web`, hoping to see the NVDL, but I would only get the Guidelines out of that, and sometimes even only the "en" Guidelines. But we need the NVDL badly, because without it, we can't work on a proposal for a new attribute class. 

I have also tried invoking the `dist` sub-targets separately (so, e.g., `dist-source dist-schema`), but it looks like `dist` is an all or nothing target, or at least I don't know how to successfully invoke just some of its targets. And it looks like there is no negotiating with `dist` wrt archives -- they are produced automatically (and delay the processing hugely).

Any hints you could push my way, please?

Thanks in advance!

**Comment by sydb** [comment, marked as answer]
Hi @bansp! Sorry, I did not even notice this posting until @HelenaSabel gave a poke. (Thanks @HelenaSabel!)

I cannot speak about what to do within the _Jenkins_ environment, as I have not ever done so myself. (But probably will in a month or two — we are getting our own CI server! Yay!) But here are some hints you may find helpful, not necessarily in an intelligent order.

1. `p5.nvdl` and `p5valid.nvdl` are not (to my knowledge) _built_, they just _exist_. (All the Makefile does, besides attempt to validate against them, is to copy them over to a release directory.)
2.   I think clean, validate, html-web, test, & exemplars is exactly the set of targets you probably want to execute.
3. VCS=git is not really necessary, git is the default. (No harm, either.)
4. I do not think INJENKINS=true is strictly necessary when VCS=git, but I would leave it anyway. (Can’t hurt, might help, especially if someone changes the Makefile in the future expecting you to have it.)
5. Since you do not need the README files or the _Guidelines_ in multiple languages, my instinct is to steer clear of the dist* targets. If there is something you really need in there, though (other than *.nvdl, which can just be copied), let me know, I can poke at this further.

In case it is remotely helpful, what I typically do _outside_ the Jenkins environment is either
~~~
$ cd /path/to/TEI_repo/P5
$ time make clean validate html-web test exemplars XSL=/path/to/Stylesheets_repo
~~~
or, in order to execute within the TEIC “jenkins” Docker container
~~~
$ time sudo docker run --rm -v /path/to/TEI_repo:/TEI -v /path/to/Stylesheets_repo:/stylesheet -w /TEI/P5 -it --entrypoint=make teic/jenkins clean validate html-web test exemplars XSL=/stylesheet
~~~  
I have tried adding `--jobs=4 -Oline` (or even `--jobs=$(nproc 2>/dev/null || echo 1) -Oline`), but that has failed dramatically. (I think, but am not sure, that what happens is one of the parallel jobs wants to use a directory or p5.xml that another job is supposed to have already built, but hasn’t yet.)

**Reply by bansp** [reply]
Dear Syd, many thanks for these suggestions. I am definitely going to try them when I'm past my (hopefully) temporary time congestion... :-)

**Comment by bansp** [comment]
Dear Syd, thanks again for all the hints -- with some tweaking, I was eventually able to reduce the processing time at the Paderborn Jenkins by a factor of 10(!). 

Except I am not sure what exactly I've given up in order to get there, so let me throw in some unstructured remarks from the battlefield, in order to share and perhaps elicit a comment _some time in the future_ (nothing here is critical).

1. A side remark to my side remarks is on the result of `make clean` -- shouldn't it remove the following?
```
$ git status
On branch lingsig-att.linguistic.dependency
Your branch is up-to-date with 'origin/lingsig-att.linguistic.dependency'.

Untracked files:
  (use "git add <file>..." to include in what will be committed)
        Exemplars/detest_odd_schematron.svrl
        Exemplars/isofs.svrl
        Exemplars/tei_all.svrl
        Exemplars/tei_allPlus.svrl
        Exemplars/tei_bare.svrl
        Exemplars/tei_basic.svrl
        Exemplars/tei_corpus.svrl
        Exemplars/tei_customization.svrl
        Exemplars/tei_drama.svrl
        Exemplars/tei_enrich.svrl
        Exemplars/tei_its.svrl
        Exemplars/tei_jtei.svrl
        Exemplars/tei_lite.svrl
        Exemplars/tei_math.svrl
        Exemplars/tei_minimal.svrl
        Exemplars/tei_ms.svrl
        Exemplars/tei_odds.svrl
        Exemplars/tei_simplePrint.svrl
        Exemplars/tei_speech.svrl
        Exemplars/tei_svg.svrl
        Exemplars/tei_tite.svrl
        Exemplars/tei_xinclude.svrl
        Test/detest_odd_schematron.log
        Test/detest_odd_schematron.svrl
        Test/detest_odd_schematron.txt
        Test/detest_xml_relaxng.log
        Test/detest_xml_schematron.log
        Test/expected-results/detest_odd_schematron.txt
        p5odds.sch
```

-- is that a to-do, a fix request for the `Makefile`?

2. The command I used locally is `make XSL=/usr/share/xml/tei/stylesheet GOOGLEANALYTICS=false clean default` -- in order to not only eliminate the archiving steps but also to reduce the number of language versions produced when I only need 'en' for testing. Not sure if the google option helps to eliminate overhead, haven't checked.

2.1. The `default` target expands into `validate exemplars test html-web`, and I will go back to using the expansion in the Jenkins config, for greater clarity. (These targets are exactly the four suggested by Syd above.)

3. Preparing for the local simulation of the Jenkins integration, I visited issue #2455 and shared a few remarks -- in case someone would like to install `rnv` for the sake of making TEI, without a docker (docker may be a more sensible path), and if their level of cluelessness approaches mine.

4. One thing that I find worrying among the warnings that I see is:
```
Validate using xmllint for DTD
  xmllint --noout --dtdvalid tei_basic.dtd tei_basic.tei

  tei_basic.dtd:1290: parser warning : PEReference: %tei_att.linguistic.dependency.attributes; not found
   %tei_att.linguistic.attribute.join;'> 
                                       ^
```

-- I wonder if that is a tei_basic issue or an expected effect? And what in the message above is "not found", the `%tei_att.linguistic.dependency.attributes;` (likely, I guess) or ` %tei_att.linguistic.attribute.join;` (unlikely, I hope)?

(These are not all issues that come to mind, but I have just broken the LingSIG build, so I'll stop for now.)
